src/balancerd/src/lib.rs

```diff
@@ -547,6 +561,41 @@ impl mz_server_core::Server for PgwireBalancer {
     }
 }
 
+async fn cancel_request(conn_id: u32, secret_key: u32, cancellation_resolver: &str) {
+    let addr = cancellation_resolver.replace("{}", &conn_id_org_uuid(conn_id));
+    let ips = match tokio::net::lookup_host(&addr).await {
+        Ok(ips) => ips,
+        Err(err) => {
+            error!("{addr} failed resolution: {err}");
+            return;
+        }
+    };
+    let mut buf = BytesMut::with_capacity(16);
+    let msg = FrontendStartupMessage::CancelRequest {
+        conn_id,
+        secret_key,
+    };
+    msg.encode(&mut buf).expect("must encode");
+    // TODO: Is there a way to not use an Arc here by convincing rust that buf will outlive the
+    // spawn? Will awaiting the JoinHandle work?
+    let buf = Arc::new(buf);
```

Comment:
I think the [freeze](https://docs.rs/bytes/1.5.0/bytes/struct.BytesMut.html#method.freeze) method is what you want, and then you can clone the returned `Bytes` like you are with the `Arc` on line 584

Reply:
Neat!